### PR TITLE
fix: `column.editor` and `gridOptions.editorFactory` type changed

### PR DESCRIPTION
**Pull request by manfredwippel**
editor and edtitor factory do not return an Editor Object but a typeof Editor to be constructed with new (...) Otherwise it is impossible to implement an Editor Factory in Typescript

### CODE REVIEW

File: src/slick.grid.ts [hunk 1]
```diff
@@ -5931,29 +5933,35 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
       Utils.emptyElement(this.activeCellNode);
     }
 
     let metadata = (this.data as CustomDataView<TData>)?.getItemMetadata?.(this.activeRow);
     metadata = metadata?.columns as any;
     const columnMetaData = metadata && (metadata[columnDef.id as keyof ItemMetadata] || (metadata as any)[this.activeCell]);
 
-    this.currentEditor = new useEditor({
+    //the editor must be constructable
+    if (typeof useEditor !== 'function') {
```
Comment (ghiscoding) [inline src/slick.grid.ts:5948]: when would that ever happen?

**Review by ghiscoding — CHANGES_REQUESTED** [review]

File: src/models/editor.interface.ts [hunk 2]
```diff
@@ -86,3 +87,10 @@ export interface Editor {
    */
   validate: (targetElm?: HTMLElement, options?: any) => EditorValidationResult;
 }
+
+export type EditorConstructor = {
+  new <TData = any, C extends Column<TData> = Column<TData>, O extends GridOption<C> = GridOption<C>>(args?: EditorArguments<TData, C, O>): Editor;
+
+  /** static flag used in makeActiveCellEditable*/
```
[the remark this answers is not in the document]
Comment (ghiscoding) [reply src/models/editor.interface.ts:94]: actually why do you add `suppressClearOnEdit` in here? It's already in `Editor` and it should be the only place it exists, there's no need to duplicate it here

File: src/models/column.interface.ts [hunk 3]
```diff
@@ -84,7 +85,7 @@ export interface Column<TData = any> {
   disableTooltip?: boolean;
 
   /** Any inline editor function that implements Editor for the cell value or ColumnEditor */
-  editor?: Editor | { model?: Editor; } | null;
+  editor?: Editor /*| { model?: Editor; }*/ | EditorConstructor | null;
```
[the remark this answers is not in the document]
Comment (manfredwippel) [reply src/models/column.interface.ts:88]: where were no references in the project to moel.Editor, so I thought this was a leftover

File: src/models/editor.interface.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (manfredwippel) [reply src/models/editor.interface.ts:94]: eslint is running automatically while typing - I am using Visual Studio
I can check why i did not get any complaints concerning the spaces

it is becaus in getEditor(..) it is used as static flag for useEditor?.suppressClearOnEdit
before the Editor object is created.

I think getEditor or the editor property of Column is never an Editor object but typeof Editor to create one later
Thats why I added the EditorConstructor.

File: src/slick.grid.ts [hunk 4]
```diff
@@ -5931,29 +5933,35 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
...
+    const editorArgs: EditorArguments<TData, C, O> = {
       grid: this,
       gridPosition: this.absBox(this._container),
       position: this.absBox(this.activeCellNode),
       container: this.activeCellNode,
       column: columnDef,
       columnMetaData,
       item: item || {},
-      event: e,
+      event: e as Event,
```
[the remark this answers is not in the document]
Comment (manfredwippel) [reply src/slick.grid.ts:5959]: Typescript requires narrowing down to one type, because EditorArguments.event is only of type Event.
It wont make a diffenece in the generated output.

File: src/slick.grid.ts (hunk 1 above)
Comment (manfredwippel) [reply src/slick.grid.ts:5948]: I had to narrow down the type for the typescript compiler, because currentEditor could be an Editor object without new(..) constructor when the Editor is provided via parameter.
maybe there is a better way to solve this, but I could not find one.

File: src/models/editor.interface.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ghiscoding) [reply src/models/editor.interface.ts:94]: > it is becaus in getEditor(..) it is used as static flag for useEditor?.suppressClearOnEdit
before the Editor object is created.

then maybe better to remove it from `Editor` then, I only want to keep 1 prop not multiple

File: src/models/column.interface.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ghiscoding) [reply src/models/column.interface.ts:88]: no it's for my other libs usage only, so you can safely remove it from in here because like I said I no longer depend on this project for my own libs (like Slickgrid-Universal and others). So just remove the commented code please

File: src/slick.grid.ts [hunk 5]
```diff
@@ -3841,19 +3843,19 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
       this._options.defaultFormatter) as Formatter;
   }
 
-  protected getEditor(row: number, cell: number): Editor | undefined {
+  protected getEditor(row: number, cell: number): Editor | EditorConstructor | null | undefined {
```
[the remark this answers is not in the document]
Comment (manfredwippel) [reply src/slick.grid.ts:3846]: This would also require to change the type of Column.editor and ColumnMetadata.editor
This invalidates a lot of code. I leave this for now.

File: src/slick.grid.ts [hunk 6]
```diff
@@ -2448,10 +2448,10 @@ export class SlickGrid<TData = any, C extends Column<TData> = Column<TData>, O e
       for (i = 0; i < cssRules.length; i++) {
         const selector = cssRules[i].selectorText;
         if (matches = /\.l\d+/.exec(selector)) {
-          columnIdx = parseInt(matches[0].substr(2, matches[0].length - 2), 10);
+          columnIdx = parseInt(matches[0].slice(2), 10);
           this.columnCssRulesL[columnIdx] = cssRules[i];
         } else if (matches = /\.r\d+/.exec(selector)) {
-          columnIdx = parseInt(matches[0].substr(2, matches[0].length - 2), 10);
+          columnIdx = parseInt(matches[0].slice(2), 10);
```
Comment (ghiscoding) [inline src/slick.grid.ts:2454]: why are you replacing `substr` with `slice`? I'm not sure if this is really safe to use and even if it is, this change should **not** be part of this PR since this is unrelated to your original issue. PR should always focus on 1 thing not multiple. Also please explain the reason to use slice before creating another PR

Comment (manfredwippel) [reply src/slick.grid.ts:2454]: string.prototype.substr is deprecated: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr
string.prototype.slice is also supportet on all browsers including IE.
https://caniuse.com/?search=slice

I did not touch other deprecated properties like Event.wich, Window.pageYOffset and Window.pageXOffset, because of possible IE problems.

Comment (ghiscoding) [reply src/slick.grid.ts]: well that's the thing about native JS, even if it's deprecated they will probably never remove it. Why? Because as I often heard from NodeJS core developers, they can't break the web, so a website that worked 20 years ago will still work 20 years from now... Same goes for `e.keyCode` (Event) and other deprecated code that is used across SlickGrid, we could replace it in the future but there's really no rush in doing so without a lot of testing

File: src/models/editor.interface.ts [hunk 7]
```diff
@@ -1,6 +1,7 @@
 import type { EditorArguments } from './editorArguments.interface';
 import type { EditorValidationResult } from './editorValidationResult.interface';
-
+import type { GridOption } from '../models/gridOption.interface';
+import type { Column } from '../models/column.interface';
```
Comment (ghiscoding) [inline src/models/editor.interface.ts:4]: you are already inside the `/models` folder, there's no need to get out and get back in on your import path. 

Also now that you have a few imports, you can probably use the barrel index import instead, I typically do that when having a few

```suggestion
import type { Column, EditorArguments, EditorValidationResult, GridOption } from './index';
```

File: src/models/editorArguments.interface.ts [hunk 8]
```diff
@@ -2,8 +2,9 @@
 import { Column, ElementPosition, PositionMethod } from './index';
 import type { SlickDataView } from '../slick.dataview';
 import type { SlickGrid } from '../slick.grid';
+import type { GridOption } from '../models/gridOption.interface';
```
Comment (ghiscoding) [inline src/models/editorArguments.interface.ts:5]: similar as above, you are already inside `models/` folder, no need to get out and get back in. However, you should remove this line and instead modify the import on line 2, so please modify line 2 as suggested below

```suggestion
import type { Column, ElementPosition, GridOption, PositionMethod } from './index';
```
also notice that I forgot to add the `type` so please make sure you add it. It is better for import and build perf

**Comment by ghiscoding** [comment]
@manfredwippel I found 2 more lines to update, if you could please fix them, we could then go ahead with the merge. Thanks

File: src/models/editor.interface.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import type { EditorArguments } from './editorArguments.interface';
 import type { EditorValidationResult } from './editorValidationResult.interface';
-
+import type { GridOption } from '../models/gridOption.interface';
```
Comment (ghiscoding) [inline src/models/editor.interface.ts:3]: after the CI ran I see that the build is failing, 1 of the reason is that you have duplicate import, you can remove line 3 since `GridOption` is imported in line 4 as well so you have duplicate imports.

**Comment by ghiscoding** [comment]
@manfredwippel the build is failing in the UI, I have identified at least 1 duplicate import as shown above. The CI runs the `build:prod` script as shown below, so please make sure you run it too `npm run build:prod`

https://github.com/6pac/SlickGrid/blob/61e73db0027d10714a8b4ca8f19f4ccb37269301/package.json#L58

File: src/models/editorArguments.interface.ts [hunk 10]
```diff
@@ -2,8 +2,9 @@
 import { Column, ElementPosition, PositionMethod } from './index';
 import type { SlickDataView } from '../slick.dataview';
 import type { SlickGrid } from '../slick.grid';
+import type { Column, ElementPosition, GridOption, PositionMethod } from './index';
```
Comment (ghiscoding) [inline src/models/editorArguments.interface.ts:5]: more duplicate imports since line 2 and 5 are nearly the same, line 5 seems better to keep since it adds the missing `type` which is good

File: src/models/editor.interface.ts [hunk 11]
```diff
@@ -77,12 +78,18 @@ export interface Editor {
...
+  
   /**
    * Validate user input and return the result along with the validation message.
    * if the input is valid then the validation result output would be returning { valid:true, msg:null }
    * The first argument "targetElm" is ONLY used internally by the Composite Editor in most cases you want to make this null or undefined
    */
   validate: (targetElm?: HTMLElement, options?: any) => EditorValidationResult;
 }
+
+export type EditorConstructor = {
```
Comment (ghiscoding) [inline src/models/editor.interface.ts:90]: `EditorArguments` is not Generic, this will also fail the build

File: src/models/editorArguments.interface.ts (hunk 10 above)
Comment (manfredwippel) [reply src/models/editorArguments.interface.ts:3]: I am sorry - I did not run a build since I accepted your changes in the browser only and did not sync.

Comment (ghiscoding) [reply src/models/editorArguments.interface.ts:3]: np

**Comment by ghiscoding** [comment]
ok it looks good now, let's merge. I'm not sure when I'll push a new release but sometime this week  I guess, if not please remind me

**Comment by ghiscoding** [comment]
@manfredwippel I have replicated your PR into my lib Slickgrid-Universal and I found a small problem, you've assigned `EditorArguments` as possible undefined to the `EditorConstructor` but that's technically impossible because without it, the Editor simply won't work. So would you like to create another PR to resolve this issue?

https://github.com/6pac/SlickGrid/blob/e02cf641a393528ae366edacf86b8f3eec6ef8a5/src/models/editor.interface.ts#L87-L88

**Comment by ghiscoding** [comment]
now available in v5.8.2, thanks for the contribution
